55_revocation_actions_without_python.md

```diff
@@ -0,0 +1,311 @@
+will need to be a means to execute actions without Python runtime
+installation.
+
+### Goals
+
+<!--
+List the specific goals of the enhancement.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+ * Arbitrary scripts/executable can be used as a revocation action
+
+### Non-Goals
+
+<!--
+What is out of scope for this enhancement?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+ * Secure execution of revocation actions (while it is desired), such as using sandbox and seccomp
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The actions defined in the configuration file (or tenant-provided
+action list) can be any type of script/executable.
+
+When Keylime agent receives a revocation message, it stores the JSON
+payload in a file and invokes revocation actions with the file as a
+command-line argument.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this enhancement is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+* The agent is configured with revocation notification enabled
+* The agent is also configured with non-Python actions (pre-installed on the system) set up in the `revocations_actions` configuration option
+* The verifier notifies a revocation
+* The agent will invoke the actions with the revocation message stored in a file
+
+#### Story 2
+
+* The agent is configured with revocation notification enabled
+* The tenant sends non-Python actions as part of the initial payload, as well as the `action_list` file listing those actions
+* The verifier notifies a revocation
+* The agent will invoke the actions with the revocation message stored in a file
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+enhancement ecosystem.
+
+How will security be reviewed and by whom?
+-->
+
+This mechanism makes it a little easier for the attacker to execute
+commands on the system, though the same thing is already possible by
+injecting custom Python actions.  To keep the risk minimal, this
+proposal suggest mandating that the pre-installed actions are
+installed in a fixed/immutable directory.
```

Comment:
The Python modules are already somewhat fixed by being in `revocation_actions` Python module, correct?
I agree that this is a good idea to enforce the storage location of the local action scripts.

Reply:
We might consider adding the ability to turn off actions sent as payloads. For systems that are really locked down, where the revocation actions are known ahead of time, having the ability to refuse payload actions makes it more secure. Not sure if it makes sense to add to this proposal or have as something separate, but it seems small enough to include to me.

Reply:
As to the storage location, I can see the benefit of using something under `/usr/libexec` vs `/var/lib` since the latter is usually ignored during attestation because it has files that change. If we're talking pre-installed actions, those probably belong outside of `/var/lib`

Reply:
Thanks for the suggestions; I've added a new paragraph to describe options for those (`revocation_actions_dir` for the location, `allow_payload_revocation_actions` for opting out payload actions).